BUG: Optimize: NewtonCG min crashes with xtol=0

TimButters opened the pull request:
The minimize function crashes if `xtol` is set to `0`. This is because the initial value of a loop variable is `2 * xtol`, which is intended to ensure the loop runs at least once. However, if `xtol = 0` then the loop never runs, a variable isn't instantiated that is used outside of the loop, and an `UnboundLocalError` is raised.

To fix this, the initial value of the loop variable is set to the max float value.

<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
Closes #20214 

#### What does this implement/fix?
UnboundLocalError is thrown if `xtol` is set to 0 for the `newton-cg` method in `minimize`.

#### Additional information

TimButters committed:
BUG: Optimize: NewtonCG min crashes with xtol=0

The minimize function crashes if `xtol` is set to `0`. This is
because the initial value of a loop variable is `2 * xtol`, which
is intended to ensure the loop runs at least once. However, if
`xtol = 0` then the loop never runs, a variable isn't instantiated
that is used outside of the loop, and an `UnboundLocalError` is
raised.

To fix this, the initial value of the loop variable is set to the
max float value.

lucascolley reviewed (approved):
LGTM!

tylerjereddy reviewed (commented):
I'll let the optimize regulars chime in on whether a regression test is sensible to add, but this slightly modified version of the original reproducer fails before/passes after and seems fast:

```diff
--- a/scipy/optimize/tests/test_optimize.py
+++ b/scipy/optimize/tests/test_optimize.py
@@ -3120,3 +3120,19 @@ def test_sparse_hessian(method, sparse_type):
     assert res_dense.nfev == res_sparse.nfev
     assert res_dense.njev == res_sparse.njev
     assert res_dense.nhev == res_sparse.nhev
+
+
+def test_gh_20214():
+    def cosine(x):
+        f = np.cos(x[0])
+        return f
+
+    def jac(x):
+        ret = -np.sin(x)
+        return ret
+
+    result = optimize.minimize(cosine,
+                               x0=[0.1],
+                               jac=jac,
+                               method="newton-cg",
+                               options=dict(xtol=0, maxiter=2))
```

h-vetinari reviewed (dismissed):
At least one error from #20214 occurs elsewhere, namely `update` is undefined in this branch:
https://github.com/scipy/scipy/blob/cd60e810c17442f71ab7f7788eccd1013fc265f1/scipy/optimize/_optimize.py#L2118-L2120

Looking at https://github.com/scipy/scipy/commit/c8c78d7171c074b8e5b97c57944b827fd13aa979, this commit failed to rename that instance of `update` to `update_l1norm`. CC @lorentzenchr 

We need a test that exercises this else-branch in any case. I haven't checked if @tylerjereddy's case does that, but to reach the `else:` clause of the `while:` loop, we must achieve `update_l1norm <= xtol` without having triggered any of the early `return` statements in the while body.

h-vetinari commented on scipy/optimize/_optimize.py:
```diff
@@ -2025,7 +2025,7 @@ def terminate(warnflag, msg):
 
     xtol = len(x0) * avextol
-    update_l1norm = 2 * xtol
+    update_l1norm = np.finfo(float).max
```
I cannot imagine this being the right solution to be honest. This means the `update_l1norm > xtol` clause of the while-loop will ~always be true, and completely ignores that this should have something to do with the initial norm.

The _very_ first thing here should be to fix the `UnboundLocalError`, not hide it from happening.

j-bowhay replied:
I agree I think the correct fix is updating `update` -> `update_l1norm` on line 2119

TimButters replied:
There are two related issues that I was trying to address with this:

1. UnboundLocalError if xtol=0
2. The function will not succeed if xtol=0, even if an exact solution can be found (as is the case with the example given in the bug (https://github.com/scipy/scipy/issues/20214)

I completely agree that the most important issue to fix is the Unbound Error, but if we only make the change `update` -> `update_l1norm` we will still fail to find the valid solution for the presented test case.

I could well be missing something, but I don't think the initial value _does_ need to be related to the initial norm. The variable is only used to enter the loop (which is guaranteed for any value that isn't zero). It is then updated at the end of the loop, it only appears to be used to track the exit condition of the while.

My idea with using the max float as the initial value is that it retains the guarantee of running at least once, and also stops any issues with people passing in extremely large values for `xtol` that could overflow.

I'd certainly welcome your thoughts @h-vetinari  and @j-bowhay - I'm very new to SciPy so could well be misunderstanding how this works.


Whatever the right solution, I should have added a test! I'm happy to add one in as suggested.

j-bowhay replied:
Ok yes I see your point the initial assignment is effectively only to enter the loop (basically a do while). However I am not sure the change is correct because previously `2 * xtol` is always greater than `xtol` whereas now this not the case. Prehaps it should initially just be set to `np.inf`.

Regardless, the mistake on 2119 should be fixed and it would be good add the test.

j-bowhay replied:
Ha sorry a very silly mistake from me, I miss read this as min not max, I think this should be fine

TimButters replied:
Thanks. The latest revision also includes a test and corrects the mistake on line 2119, so I think that covers everything.

h-vetinari replied:
There's one thing that is being changed here, which is that previously, `update_l1norm` maintained the shape of `x0`, and in particular, `xtol` will _also_ have the shape of `x0`.

I don't think this ever worked for `len(x0)>1` though, as `while update_l1norm > xtol:` would then run into:
```
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

I don't know the code here well enough, but this stands out to me...

TimButters committed:
BUG: Optimize: UnboundLocalError for xtol=0 NCG

Corrected error from previous rework when switching to
`np.linalg.norm` that introduced the possibility of the unbound
variable error.

TimButters committed:
TST: Optimize: NewtonCF Minimize when xtol=0

Regression test for gh-20214 which addressed an issue in the
Newton CG method which would raise an UnboundLocalVariable error
if `xtol` was set to zero. It also ensures the method finds the
solution in this case.

j-bowhay reviewed (approved):
Thanks @TimButters lgtm but I will wait to see if others are happy before merging

h-vetinari commented on scipy/optimize/_optimize.py:
```diff
@@ -2116,7 +2116,7 @@ def terminate(warnflag, msg):
 
     else:
-        if np.isnan(old_fval) or np.isnan(update).any():
+        if np.isnan(old_fval) or np.isnan(update_l1norm):
```
... not least because you are also removing the call to `.any()` here.

TimButters replied:
Happy to add the `.any()` back in, but the norm is always going to be a scalar so it's unnecessary. The norm will always be `nan` if any of the elements of the vector (`update`) are `nan`.

I'm not sure about the shape side of this. It looks like previously `update_l1norm` was initialised to `2 * xtol`, and `xtol` is `len(x0) * avextol`. With `len(x0)` and `avextol` being scalars I think this has always just been a float value?

h-vetinari reviewed (commented):
I'm still unsure about the right initial value for `update_l1norm`, though I can see the argument that it's being updated at the end of the loop. Still, the interplay with the shape of `x` is a mystery to me. The code seems very brittle w.r.t. to putting in `x0=[1, 2]` or something like that.

Unless we can figure this out quickly _and_ comprehensively, I'd prefer an immediate fix for the unboundname error to go into 1.13, and the rest to be fixed for 1.14.

lorentzenchr commented on scipy/optimize/_optimize.py:
```diff
@@ -2116,7 +2116,7 @@ def terminate(warnflag, msg):
 
     else:
-        if np.isnan(old_fval) or np.isnan(update).any():
+        if np.isnan(old_fval) or np.isnan(update_l1norm):
```
Is it guaranteed that update_l1norm is nan whenever there‘s a nan in update?

TimButters replied:
Yes, `np.linalg.norm` will always return `nan` if one or more elements of the vector is `nan`.

lorentzenchr commented on scipy/optimize/tests/test_optimize.py:
```diff
@@ -507,6 +507,23 @@ def test_ncg_negative_maxiter(self):
+                                   jac=jac,
+                                   method="newton-cg",
+                                   options=dict(xtol=xtol))
+        assert result.status == 0
```
Could we also test the value of the obtained solution?

TimButters replied:
Good point, I've added that to the test.

lorentzenchr commented:
Is there a test for the „nan-path“?

TimButters committed:
MAINT: Add comment to explain initial value

Comment added to explain the purpose of setting `update_l1norm` to the max float value - this is to ensure the preceding while loop executes at least once.

Co-authored-by: Christian Lorentzen <lorentzen.ch@gmail.com>

TimButters committed:
TST: Test solution value as well as success

This is for the `xtol=0` test in the Newton CG minimize method.

TimButters commented:
> Is there a test for the „nan-path“?

@lorentzenchr yes, there is an existing test `test_nan_values` that checks this branch.
Line 1494 in `test_optimize.py`.

TimButters committed:
TST: Explicitly extract val from array

TimButters commented:
@j-bowhay @lucascolley I think we have consensus now, would one of you be happy to merge this?

lucascolley reviewed (approved):
thanks @TimButters , I'll give it a few days for any more comments before merging.

As a note for the future, I'd recommend that you leave resolving conversations to maintainers - a thorough reviewer will probably want to check any conversations that were resolved by the author, so it is easier to leave them as is.

TimButters commented:
Thanks for the feedback @lucascolley, I thought I was doing the right thing there. I’ll leave them open in the future.

lucascolley commented:
> I'll give it a few days for any more comments before merging.

Here we are, thanks @TimButters !

j-bowhay commented:
Thanks @TimButters a great couple of first contributions, keep them coming!
